WIP: Topics/testing

zakalawe opened the pull request:
Current state of the test-harness and driver, for feedback and discusson. @Marcus10110

zakalawe committed:
Settings data defined

This won’t build yet, since I didn’t include commented-out code
to compile the rest of the plugin.

zakalawe committed:
Basic CMake build system

Needs Linux, Windows support and installation path handling still.

zakalawe committed:
Data-generator work in progress

zakalawe committed:
WIP on analysis stage

zakalawe committed:
Use RGB struct in analysis, results

zakalawe committed:
Use RGB struct in analysis, results

zakalawe committed:
Revised settings after discussion with Mark

zakalawe committed:
Data generation fixes

zakalawe committed:
Settings bug fix in timing read-out

zakalawe committed:
Settings MSVC float -> int fix

zakalawe committed:
MSVC / Linux Cmake build from Mark

zakalawe committed:
Add RGB helper struct

Use this to deal with different color layouts and bit sizes in
a cleaner way.

zakalawe committed:
Synchronise to a RESET pulse on analysis

Avoids a warning about the first RESET pulse, since we scan the stream
forwards for the first RESET we see and begin analysis from that point.

zakalawe committed:
Settings / data-generator changes based on review

- define timings with tolerances, min, max
- use BitExtractor in the data-generator
- assume the data signal defaults to low, not high
- define some helper structs for managing timings with tolerances
- generate both high-sped and low-speed data

zakalawe committed:
Analysis updates based on review feedback

zakalawe committed:
Results / bubble text according to spec.

zakalawe committed:
Fix missing data for the LPD1886

zakalawe committed:
Data-generator: make high-speed mode conditional

zakalawe committed:
Analysis: better speed mode detection

This gives reliable classification of mixed-speed simulated data.

zakalawe committed:
Results CSV export generation

zakalawe committed:
Portability fixed from Mark’s review

zakalawe committed:
Implement GenerateFrameTabularText to spec.

zakalawe committed:
Add AStyle support

zakalawe committed:
AStyle changes

zakalawe committed:
Remove stray comment :)

zakalawe committed:
Hacking on test-harness for analyzers

zakalawe committed:
More work on the test harness. MockChannelData
can be defined and the AnalyzerChannelData API is implemented, so
close to being able to run the worker thread.

zakalawe committed:
Even more hacking on the test-harness

Close to working now, frames and packets are analysed and collected.

zakalawe committed:
Simple analysis test works!

zakalawe committed:
Expanding test coverage

- test synchronise mid-way through the data
- force the analyzer to resync due to bad data

zakalawe committed:
Add simple bubble text testing

- this also uncovered a bug in the analyzer, and
a problem with string formatting

zakalawe commented:
@Marcus10110 pushed some more commits, tests are running now so it's a good time to get feedback and as in-depth a code reivew as you can do.

Marcus10110 commented on AnalyzerTestHarness/SettingsStubs.cpp:
```diff
@@ -0,0 +1,247 @@
+#include "AnalyzerSettings.h"
+#include "AnalyzerSettingInterface.h"
+
+#include <cmath>
```
```
#include <cassert>
#include <algorithm>
```

zakalawe replied:
Done

Marcus10110 commented on AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp:
```diff
@@ -0,0 +1,88 @@
+
+BitState SimulationChannelDescriptor::GetCurrentBitState()
+{
+
```
MSVC requires this to return a value.

zakalawe replied:
Done with a stub for now, will be expanded when we test simulated data.

Marcus10110 commented on AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp:
```diff
@@ -0,0 +1,88 @@
+}
+
+U64 SimulationChannelDescriptor::GetCurrentSampleNumber()
+{
```
MSVC requires this to return a value.

zakalawe replied:
Ditto.

Marcus10110 commented on AnalyzerTestHarness/HelperStubs.cpp:
```diff
@@ -0,0 +1,255 @@
+
+U64 AnalyzerHelpers::AdjustSimulationTargetSample(U64 target_sample, U32 sample_rate, U32 simulation_sample_rate)
+{
+
```
MSVC requires this to return a value.

zakalawe replied:
Copied the (two line) impl from Logic's AnalyzerHelpers.cpp

Marcus10110 commented on AnalyzerTestHarness/HelperStubs.cpp:
```diff
@@ -0,0 +1,255 @@
+ }
+
+const char* SimpleArchive::GetString()
+{
```
MSVC requires this to return a value.

zakalawe replied:
Stubbed for now, until we do real testing of settings archival

Marcus10110 commented on AnalyzerTestHarness/TestInstance.cpp:
```diff
@@ -0,0 +1,141 @@
+#include "MockSimulatedChannelDescriptor.h"
+
+#include <cassert>
+#include <iostream>
```
```
#include <cstring> //for memcpy on gcc
```

Marcus10110 replied:
this cstring is still missing, and breaking the build with gcc.

Marcus10110 commented on tests/AsyncRgbLedTestDriver.cpp:
```diff
@@ -0,0 +1,335 @@
+#include "MockSettings.h"
+
+#include <cassert>
+#include <iostream>
```
```
#include <cstring> //for strcmp
```

Marcus10110 commented on CMakeLists.txt:
```diff
@@ -44,7 +44,17 @@ target_include_directories(AsyncRgbLedAnalyzer PRIVATE ${ANALYZER_SDK_ROOT}/incl
+add_executable(AsyncRgbLedTest tests/AsyncRgbLedTestDriver.cpp ${SOURCES})
+target_link_libraries(AsyncRgbLedTest AnalyzerTestHarness)
 
+add_test(AsyncRgbLedTest ${EXECUTABLE_OUTPUT_PATH}/${name})
```
This doesn't seem to work on my machine, and I don't know where ${name} comes from. I changed this to:
```
add_test(AsyncRgbLedTest AsyncRgbLedTest)
```
and I added `enable_testing()` above. then `make test` ran successfully on Linux.

Marcus10110 reviewed (changes requested):
included are comments to restore building on gcc and MSVC.

There is one other problem on MSVC.

The following classes are marked with `LOGICAPI`:
```
AnalyzerSettingInterfaceText
AnalyzerSettingInterfaceBool
AnalyzerSettingInterfaceInteger
```
However, none of their members are implemented. MSVC gives linker errors for the destructor and the `GetType` methods on each of these. I'm not 100% sure why, but removing `LOGICAPI` from their definitions fixes the errors and allows the project to link and for the tests to run and pass.
`LOGICAPI` expands to `__declspec(dllexport)` on MSVC.
We could just implement these methods to get it to link, or find another solution.

zakalawe committed:
Add missing helpers - fixes bubble text testing

zakalawe committed:
Enable CTest in the Cmake files

zakalawe committed:
Fixes based on Mark’s review

zakalawe committed:
Test tabular text generation

zakalawe committed:
WIP on verify simulation data in tests

Reset pulse is verified for now, next step is verifying the actual
data pulses

zakalawe committed:
Testing simulation generation

We don’t examine the actual numbers yet, but validating the
structure including reset pulses and channel data

Marcus10110 commented:
just caught up with the latest changes, looks great!

zakalawe commented:
@Marcus10110 do you want me to land these changes on master? Or hold off a bit longer? Will reply to other comments via email later

Marcus10110 commented:
@zakalawe I changed the base of this PR to `topics/analysis` so that the "Files Changed" view would only reflect changes since the last PR.
Ultimately let's get everything into master, which we can fast-forward once testing is finished.

zakalawe committed:
Fix a bug discovered by the test-code!

WS2811 high-speed timing was wrong for one of the bits. Testing is
valuable :)

zakalawe committed:
Refactor common code into the test harness

Simplifying test-driver code slightly, preparing for re-use in
other analyzer tests

zakalawe committed:
WIP towards programmable test data

Restructuring test code into a helper class, so different controllers
and tolerances can be configured.

zakalawe committed:
Mode config for the WS2811

zakalawe committed:
Wider tolerance testing, more controllers

Expanding the test coverage and tolerancing, which is uncovering some
real bugs in the analyzer.

zakalawe committed:
Work on test-data accuracy

Trying to ascertain why analyser plugin reports smaller sample counts
(by one) than the mock test data. Adding some test-harness verification
code to check for this in the MockChannelData API implementation.

zakalawe committed:
Ignore file for test-harness

zakalawe committed:
Updated AsyncRgbLed ignore file

zakalawe committed:
Further test-driver improvements

More coverage of each mode, tolerance and controller chip.

zakalawe committed:
Further test-driver improvements

More coverage, unfortunately some more failures to fix now

Marcus10110 committed:
changed tolerance on WS2812B based on new source.

Marcus10110 committed:
Merge pull request #5 from saleae/topics/tolerance-update

changed tolerance on WS2812B based on new source.

zakalawe committed:
Latest analyzer SDK

zakalawe committed:
Update for test-harness API changes

Marcus10110 reviewed (changes requested):
Builds and tests pass on MacOS and Linux.
On MSVC, there are a few issues - hannah found these too and included fixes for them here: 
https://github.com/TheOneRing/AsyncRgbLedAnalyzer/commit/b829efc6d94ce51ab45d2ef41d100700b5599424
However I suggest using either the C++11 new random features to replace `drand48()`.
Also now I don't think adding NOMINMAX to the defines in cmake was necessary to fix the uses of `std::max` and `std::min`, I think that might have just needed `#include <algorithm>`
Once this builds and tests pass out of the box on all 3 platforms, we can get it merged!

zakalawe committed:
Portability: don’t rely on drand48()

zakalawe committed:
Bump to master branch of the SDK

zakalawe commented:
@Marcus10110 Hopefully this passes on Linux and Windows now? I didn't address the rpath issue however

Marcus10110 commented:
@zakalawe `#include <algorithm>` is still missing from `AsyncRgbLedTestDriver.cpp` which breaks the build on MSVC. After adding that, it builds and passes on MSVC.

zakalawe committed:
Windows build fix

Marcus10110 reviewed (approved):
woot! this builds and passes on all 3 platforms, merging to master!

zakalawe commented:
Wonderful!
